File: src/main/java/io/github/nikpivkin/OnTimeOutCallback.java
package io.github.nikpivkin;

import io.github.nikpivkin.localize.Localizer;
import io.github.nikpivkin.localize.Messages;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.temporal.ChronoUnit;
import java.util.Date;
import java.util.Optional;
import net.kyori.adventure.text.Component;
import org.bukkit.entity.Player;
import org.bukkit.event.player.PlayerKickEvent.Cause;

public sealed interface OnTimeOutCallback {

  void apply(Player player);

  final class Kick implements OnTimeOutCallback {

    private final Localizer localizer;

    public Kick(Localizer localizer) {
      this.localizer = localizer;
    }

    @Override
    public void apply(Player player) {
      player.kick(
          Component.text(
              localizer.translate(
                  player.locale(),
                  Messages.PLAYER_YOUR_TIME_IS_UP
              )
          ),
          Cause.PLUGIN
      );
    }
  }

  final class Ban implements OnTimeOutCallback {

    private final Localizer localizer;
    private final Optional<Long> blockingTimeInSeconds;

    public Ban(Localizer localizer, Long blockingTimeInSeconds) {
      this.localizer = localizer;
      this.blockingTimeInSeconds = Optional.ofNullable(blockingTimeInSeconds);
    }

    @Override
    public void apply(Player player) {
      var expires = blockingTimeInSeconds.map(
          e -> LocalDate.now().plus(e, ChronoUnit.SECONDS)
      ).map(date -> Date.from(date.atStartOfDay(ZoneId.systemDefault()).toInstant()));

      player.banPlayerFull(
          localizer.translate(
              player.locale(),
              Messages.PLAYER_YOUR_TIME_IS_UP
          ),
          expires.orElse(null), null)
      ;
    }
  }
}

File: src/main/kotlin/io/github/nikpivkin/lifetime/OnTimeOutCallback.kt
package io.github.nikpivkin.lifetime;

import io.github.nikpivkin.lifetime.localize.Localizer
import io.github.nikpivkin.lifetime.localize.Messages
import net.kyori.adventure.text.Component
import org.bukkit.entity.Player
import org.bukkit.event.player.PlayerKickEvent.Cause
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.*

sealed interface OnTimeOutCallback {

  fun apply(player: Player);

  class Kick (
      private val localizer: Localizer
  ): OnTimeOutCallback {

    override fun apply(player: Player) {
      player.kick(
          Component.text(
              localizer.translate(
                  player.locale(),
                  Messages.PLAYER_YOUR_TIME_IS_UP
              )
          ),
          Cause.PLUGIN
      );
    }
  }

  class Ban (
      private val localizer: Localizer,
      private val blockingTimeInSeconds: Optional<Long>
  ): OnTimeOutCallback {

    override fun apply(player: Player) {
      val expires = blockingTimeInSeconds
          .map { LocalDate.now().plus(it, ChronoUnit.SECONDS) }
          .map { Date.from(it.atStartOfDay(ZoneId.systemDefault()).toInstant()) }
          .orElse(null)

      player.banPlayerFull(
          localizer.translate(
              player.locale(),
              Messages.PLAYER_YOUR_TIME_IS_UP
          ),
          expires,
          null
      )
    }
  }
}
